### PR TITLE
Port bats test to go test

### PR DESCRIPTION
**Pull request by tiffanny29631**
cli.bats
- test case converted and added check for multi-repo mode

foo_corp.bats
- test case already covered by acme_test.go, removed

repo_less.bats
- test case converted into mono repo only go test

The related BATS files are removed.

### CODE REVIEW

File: e2e/testcases/cli_test.go
Comment (haiyanmeng) [inline e2e/testcases/cli_test.go:52]: Use `ControllerNamespace` from pkg/api/configsync

File: e2e/testcases/cli_test.go
Comment (haiyanmeng) [inline e2e/testcases/cli_test.go:53]: Use `MonitoringNamespace` from `pkg/metrics`

File: e2e/testcases/cli_test.go
Comment (haiyanmeng) [inline e2e/testcases/cli_test.go:54]: Use `RGControllerNamespace` from `pkg/api/configmanagement`

File: e2e/testcases/cli_test.go
Comment (haiyanmeng) [inline e2e/testcases/cli_test.go:853]: `ntopts.RequireGKE(t)`: is this required?

File: e2e/testcases/cli_test.go
Comment (haiyanmeng) [inline e2e/testcases/cli_test.go:873]: Do sth to avoid the duplication of the following statements:
```
       err = nt.Validate(..., "", &corev1.Namespace{})
		if err != nil {
			tw.Fatal(err)
		}
```

File: e2e/testcases/cli_test.go
Comment (haiyanmeng) [inline e2e/testcases/cli_test.go:904]: Do sth to avoid duplication.

File: e2e/testcases/cli_test.go
Comment (haiyanmeng) [inline e2e/testcases/cli_test.go:990]: Do sth to avoid duplication.

File: e2e/testcases/cli_test.go
Comment (haiyanmeng) [inline e2e/testcases/cli_test.go:1002]: Why not use `nt.T.Fatal` here?

File: e2e/testcases/legacy_test.go
Comment (haiyanmeng) [inline e2e/testcases/legacy_test.go:214]: `// TODO: implement nomos status in CLI, this may be a go rewrite`: this should be removed.

File: e2e/testcases/repoless_test.go
Comment (haiyanmeng) [inline e2e/testcases/repoless_test.go:33]: Put `../raw-nomos/manifests/importer_repoless.yaml` into a var since it is used at multiple places.

File: e2e/testcases/repoless_test.go
Comment (haiyanmeng) [inline e2e/testcases/repoless_test.go:68]: The logic here seems to be used in all the three tests in this file. Do sth to avoid duplication.

File: e2e/testcases/repoless_test.go
Comment (haiyanmeng) [inline e2e/testcases/repoless_test.go:29]: The three tests in this file share lots of common logic. Do sth to avoid duplication.

**Review by haiyanmeng — CHANGES_REQUESTED** [review]

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go]: Done

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go]: Using status.MultiError instead to reduce the error catch.

File: e2e/testcases/repoless_test.go
Comment (tiffanny29631) [reply e2e/testcases/repoless_test.go]: Added const for three manifest paths.

File: e2e/testcases/repoless_test.go
Comment (tiffanny29631) [reply e2e/testcases/repoless_test.go]: Split resource validate out.

File: e2e/testcases/repoless_test.go
Comment (tiffanny29631) [reply e2e/testcases/repoless_test.go:30]: Split commit sample repo out.

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go]: Done by using status.MultiError

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go:963]: We want to validate all these files exist under the specific path, and the folder structure is different when running under mono or multi repo mode.

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go:855]: The test errors out when running with kind cluster, that has trouble exporting the kubeconfig or dump the log. I got it running in GKE clusters.

Comment (haiyanmeng) [reply e2e/testcases/cli_test.go:855]: Can the old bat test run on kind clusters?

File: e2e/testcases/cli_test.go
Comment (haiyanmeng) [reply e2e/testcases/cli_test.go:963]: I see.

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go:855]: It actually could.. 

```
-  # Fixes a problem when running the tests locally
-  if [[ -z "${KUBECONFIG+x}" ]]; then
-    export KUBECONFIG="${HOME}/.kube/config"
-  fi
```
This seemed to be doing the trick but looks confusing

File: e2e/testcases/cli_test.go
Comment (nan-yu) [inline e2e/testcases/cli_test.go:897]: It generates the bugreport in the current directory. If you move it to `nt.TmpDir`, you wouldn't need to delete it in the end. The test framework will blow away the entire `nt.TmpDir` after the test.

File: e2e/testcases/cli_test.go
Comment (nan-yu) [reply e2e/testcases/cli_test.go:963]: Can we define a list of expected files and then iterate through the list using a for loop?

```
    expectedFiles := []string{
		fmt.Sprintf("%s/processed/%s/version.txt", bugReportDirName, context),
		fmt.Sprintf("%s/processed/%s/status.txt", bugReportDirName, context),
	}
	if nt.MultiRepo {
		expectedFiles = append(expectedFiles, ...)
		...
	} else {
		expectedFiles = append(expectedFiles, ...)
		...
	}

	for _, file := range expectedFiles {
		if checkFileExists(file, files) {
			nt.T.Errorf("did not find all expected files in bug report zip file")
		}
	}

```

Using the above code should get rid of the `found` flag.

**Review by nan-yu — CHANGES_REQUESTED** [review]

File: e2e/testcases/repoless_test.go
Comment (nan-yu) [inline e2e/testcases/repoless_test.go:40]: This is not needed as the [RresetMonoRepoSpec](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/e2e/nomostest/config-sync.go#L1211) will reset the policy_dir back to the initial one.

**Review by nan-yu — COMMENTED** [review]
It is okay for this PR, but for future reference, I feel like the repoless conversion should go in a separate PR.

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go:861]: Moving zip file and unzipping into nt.TmpDir

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go]: Fixed, it's because of the resource validation predicate. Now the test works with kind.

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go]: No it does not, removed as the test works on kind as well.

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go]: Agreed, removed this block.

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go]: Removed as it's duplicate.

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go]: Removed the test wrapper, just using nt.

File: e2e/testcases/repoless_test.go
Comment (tiffanny29631) [reply e2e/testcases/repoless_test.go]: Done

File: e2e/testcases/repoless_test.go
Comment (tiffanny29631) [reply e2e/testcases/repoless_test.go]: Removed Cleanup for all three test cases.

File: e2e/testcases/cli_test.go
Comment (haiyanmeng) [inline e2e/testcases/cli_test.go:51]: Use RGControllerNamespace from pkg/api/configmanagement, as commented in https://github.com/GoogleContainerTools/kpt-config-sync/pull/134#discussion_r981813531.

File: e2e/testcases/legacy_test.go
Comment (haiyanmeng) [inline e2e/testcases/legacy_test.go:174]: Since this PR converts all the remaining bats tests to golang tests, we may remove this file and all the bats test files.

File: e2e/testcases/cli_test.go
Comment (nan-yu) [reply e2e/testcases/cli_test.go:51]: KubeSystem is also declared as `metav1.NamespaceSystem`.

File: e2e/testcases/cli_test.go
Comment (nan-yu) [inline e2e/testcases/cli_test.go:854]: Why not using `nomostest.New` directly? I think it should also apply to CI jobs with `--share-test-env`.

File: e2e/testcases/cli_test.go
Comment (nan-yu) [inline e2e/testcases/cli_test.go:857]: You need to switch to the test context, instead of the current context.

You can do so by two steps:
1. use the kubectl command to get the current test context: `nt.Kubectl*config", "current-context"), which takes `nt.kubeconfigPath` as `--kubeconfig`.
2. use the kubectl command to switch to the context: `nt.MustKubectl("config", "use-context", testContext)`

Then `nomos bugreport` is guaranteed to run with the test context, instead of the default one.

File: e2e/testcases/cli_test.go
Comment (nan-yu) [inline e2e/testcases/cli_test.go:874]: You can get rid of the move command by setting the working directory when running the `nomos bugreport` command:
```
cmd := exec.Command("nomos", "bugreport")
cmd.Dir = nt.TmpDir
out, err := cmd.Output()
```

File: e2e/testcases/cli_test.go
Comment (nan-yu) [inline e2e/testcases/cli_test.go:946]: I left a command a few days ago to simplify this block: https://github.com/GoogleContainerTools/kpt-config-sync/pull/134#discussion_r988440749.

File: examples/repoless/default-ns.yaml
Comment (nan-yu) [inline examples/repoless/default-ns.yaml:18]: Just curious why do we want to change the namespace? If so, we need to update the file name as it is not the default namespace anymore,

**Review by nan-yu — CHANGES_REQUESTED** [review]

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go]: Done, const replaced by specific declarations.

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go:850]: I noticed consistent presubmit failures with `nomos` command fails at creating RestMap, and the error indicates attempts to load kubeconfig from other random test directories. The root cause points to the parallelism after several tests so using this logic but skip enableParallel.

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go]: The test seems to have no issue with recognizing context, if it is for the error mentioned in [this comment](https://github.com/GoogleContainerTools/kpt-config-sync/pull/134#discussion_r998807686). I've tested it in both GKE and Kind.

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go]: Done, setting work directory when running the `nomos bugreport`

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go:941]: Done as recommended, checkFileExists now takes a list of file to verify, and returns nil or a list of files that are not found.

File: examples/repoless/default-ns.yaml
Comment (nan-yu) [reply examples/repoless/default-ns.yaml:18]: ACK

File: e2e/testcases/legacy_test.go
Comment (haiyanmeng) [reply e2e/testcases/legacy_test.go:174]: Sounds good. Thasnks!

**Comment by haiyanmeng** [comment]
/lgtm
/hold 
Please unhold this PR after Nan approves it.

File: e2e/testcases/cli_test.go
Comment (nan-yu) [reply e2e/testcases/cli_test.go:850]: I see. The `nomos` command loads the kubeconfig from either the `KUBECONFIG` env, or the `.kube/config` file under the current user HOME dir: https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/pkg/client/restconfig/config.go#L43.

One last thing to try is to set the env variable in the test and then unsets it in `nt.T.cleanup`.

If we can use `nomostest.New` directly, then we don't need to create a Kind cluster, and install Config Sync on the Kind cluster in the periodic jobs, which will eventually avoid extending the CI running time.

Comment (tiffanny29631) [reply e2e/testcases/cli_test.go]: Done by using nt and setting KUBECONFIG before running `nomos bugreport`, thanks for help debugging!

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go]: As discussed in the mentioned thread, setting kubectl context turns out to be not relevant in this scenario.

**Review by nan-yu — APPROVED** [review]
/unhold
/lgtm

File: e2e/testcases/cli_test.go
Comment (tiffanny29631) [reply e2e/testcases/cli_test.go]: Done!

**Review by haiyanmeng — APPROVED** [review]

**Pull request merged**
